elmslie-android/src/main/java/vivid/money/elmslie/android/androidstore/AndroidStore.kt

```diff
@@ -0,0 +1,105 @@
+
+import android.os.Bundle
+import androidx.activity.ComponentActivity
+import androidx.annotation.MainThread
+import androidx.fragment.app.Fragment
+import androidx.lifecycle.AbstractSavedStateViewModelFactory
+import androidx.lifecycle.SavedStateHandle
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.ViewModelStoreOwner
+import androidx.savedstate.SavedStateRegistryOwner
+import vivid.money.elmslie.core.store.Store
+
+@MainThread
+fun <
+    Event : Any,
+    Effect : Any,
+    State : Any,
+> androidStore(
+    key: String,
+    getViewModelStoreOwner: () -> ViewModelStoreOwner,
+    getSavedStateRegistryOwner: () -> SavedStateRegistryOwner,
+    getDefaultArgs: () -> Bundle,
+    storeFactory: (SavedStateHandle) -> Store<Event, Effect, State>,
+): Lazy<Store<Event, Effect, State>> =
+    lazy(LazyThreadSafetyMode.NONE) {
+        val factory =
+            RetainedViewModelFactory(
+                stateRegistryOwner = getSavedStateRegistryOwner.invoke(),
+                defaultArgs = getDefaultArgs.invoke(),
+                storeFactory = storeFactory,
+            )
+        val provider = ViewModelProvider(getViewModelStoreOwner.invoke(), factory)
+
+        @Suppress("UNCHECKED_CAST")
+        provider.get(key, RetainedViewModel::class.java).store as Store<Event, Effect, State>
+    }
+
+@MainThread
+fun <
+    Event : Any,
+    Effect : Any,
+    State : Any,
+> Fragment.store(
+    key: String = this::class.java.canonicalName ?: this::class.java.simpleName,
+    getViewModelStoreOwner: () -> ViewModelStoreOwner = { this },
+    getSavedStateRegistryOwner: () -> SavedStateRegistryOwner = { this },
+    getDefaultArgs: () -> Bundle = { arguments ?: Bundle() },
+    storeFactory: (SavedStateHandle) -> Store<Event, Effect, State>,
+): Lazy<Store<Event, Effect, State>> =
+    androidStore(
+        storeFactory = storeFactory,
+        key = key,
+        getViewModelStoreOwner = getViewModelStoreOwner,
+        getSavedStateRegistryOwner = getSavedStateRegistryOwner,
+        getDefaultArgs = getDefaultArgs,
+    )
+
+@MainThread
+fun <
+    Event : Any,
+    Effect : Any,
+    State : Any,
+> ComponentActivity.store(
+    key: String = this::class.java.canonicalName ?: this::class.java.simpleName,
+    getViewModelStoreOwner: () -> ViewModelStoreOwner = { this },
+    getSavedStateRegistryOwner: () -> SavedStateRegistryOwner = { this },
+    getDefaultArgs: () -> Bundle = { this.intent?.extras ?: Bundle() },
+    storeFactory: (SavedStateHandle) -> Store<Event, Effect, State>,
+): Lazy<Store<Event, Effect, State>> =
+    androidStore(
+        storeFactory = storeFactory,
+        key = key,
+        getViewModelStoreOwner = getViewModelStoreOwner,
+        getSavedStateRegistryOwner = getSavedStateRegistryOwner,
+        getDefaultArgs = getDefaultArgs,
+    )
+
+private class RetainedViewModel<Event : Any, Effect : Any, State : Any>(
```

Comment:
this will be used in compose artifact, most likely, so should not be private and, maybe go by a different name (e.g. RetainedElmStore)

Reply:
as well as RetainedViewModelFactory